components/raftstore-v2/src/router/imp.rs

```diff
@@ -1,13 +1,72 @@
 // Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
 
+use std::{
+    cell::RefCell,
+    sync::{Arc, Mutex},
+};
+
 use engine_traits::{KvEngine, RaftEngine};
 use raftstore::store::{FetchedLogs, LogFetchedNotifier};
+use slog::Logger;
 
 use super::PeerMsg;
-use crate::batch::StoreRouter;
+use crate::{
+    batch::StoreRouter,
+    operation::{CachedReadDelegate, LocalReader, StoreMetaDelegate},
+    StoreMeta,
+};
 
 impl<EK: KvEngine, ER: RaftEngine> LogFetchedNotifier for StoreRouter<EK, ER> {
     fn notify(&self, region_id: u64, fetched: FetchedLogs) {
         let _ = self.force_send(region_id, PeerMsg::FetchedLogs(fetched));
     }
 }
+
+/// A router that routes messages to the raftstore
+pub struct ServerRaftStoreRouter<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    router: StoreRouter<EK, ER>,
+    local_reader: RefCell<LocalReader<EK, StoreRouter<EK, ER>>>,
```

Comment:
Why `RefCell`?

Reply:
So get_snapshot can be called on & rather than &mut. Is this not needed?

Reply:
What could be the problem if using `&mut`?

Reply:
> So get_snapshot can be called on & rather than &mut. Is this not needed?

I think either make get_snapshot use &mut or make the underlying get_snapshot accept &. To me both way is fine, the question is which one needs less code change. Using RefCell here is quite dangerous  as it's easily mis-used and cause panic.

Reply:
So, why v1 uses `RefCell`?

Reply:
You may also try to get rid of `RefCell` for v1.